source/core/Console.cpp

```diff
@@ -0,0 +1,495 @@
+		}
+
+		if (ToUnicode((UInt32)keys, 0, keyboardState, buf, 256, 0) == 1)
+			output = gcnew System::String(buf);
+
+		return output;
+	}
+
+	void ConsoleScript::AddToInput(System::String ^ input)
+	{
+		if (input != nullptr)
+			_input = _input->Insert(_input->Length - _cursorPos, input);
+	}
+
+	void ConsoleScript::RemoveCharRight()
+	{
+		if (_input->Length > 0)
+			_input = _input->Remove(_input->Length - _cursorPos, 1);
+		if (_cursorPos > 0)
+			_cursorPos--;
+	}
+
+	void ConsoleScript::RemoveCharLeft()
+	{
+		if (_input->Length > 0)
+			_input = _input->Remove(_input->Length - _cursorPos - 1, 1);
+	}
+
+	void ConsoleScript::ClearInput()
+	{
+		_input = "";
+		_cursorPos = 0;
+	}
+
+	Assembly^ ConsoleScript::CompileInput()
+	{
+		String ^inputStr = String::Format(CompileTemplate, _input);
+
+		_compiler = gcnew Microsoft::CSharp::CSharpCodeProvider();
+		_compilerOptions = gcnew CodeDom::Compiler::CompilerParameters();
+		_compilerOptions->GenerateInMemory = true;
+		_compilerOptions->IncludeDebugInformation = true;
+		_compilerOptions->ReferencedAssemblies->Add("System.dll");
+		_compilerOptions->ReferencedAssemblies->Add(Script::typeid->Assembly->Location);
+
+		CodeDom::Compiler::CompilerResults ^compilerResult = _compiler->CompileAssemblyFromSource(_compilerOptions, inputStr);
+
+		if (!compilerResult->Errors->HasErrors)
+		{
+			return compilerResult->CompiledAssembly;
+		}
+		else
+		{
+			Error("Couldn't compile input-string");
+
+			StringBuilder ^errors = gcnew StringBuilder();
+
+			for (int i = 0; i < compilerResult->Errors->Count; ++i)
+			{
+				errors->Append("   at line ");
+				errors->Append(compilerResult->Errors->default[i]->Line);
+				errors->Append(": ");
+				errors->Append(compilerResult->Errors->default[i]->ErrorText);
+
+				if (i < compilerResult->Errors->Count - 1)
+				{
+					errors->AppendLine();
+				}
+			}
+
+			Error(errors->ToString());
+			return nullptr;
+		}
+	}
+
+	void ConsoleScript::ExecuteInput()
+	{
+		if (_input == "")
+			return;
+		if (_compilerTask != nullptr)
+		{
+			Error("Can't compile input - Compiler is busy");
+			return;
+		}
+		_compilerTask = Task::Factory->StartNew(gcnew Func<Assembly^>(this, &ConsoleScript::CompileInput));
+	}
+	void ConsoleScript::MoveCursorRight()
+	{
+		if (_cursorPos > 0)
+			_cursorPos--;
+	}
+	void ConsoleScript::MoveCursorLeft()
+	{
+		if (_cursorPos < _input->Length)
+			_cursorPos++;
+	}
+
+	void ConsoleScript::DrawRect(int x, int y, int width, int height, Color color)
+	{
+		const float w = static_cast<float>(width) / WIDTH;
+		const float h = static_cast<float>(height) / HEIGHT;
+		float xNew = ((static_cast<float>(x)) / WIDTH) + w * 0.5f;
+		float yNew = ((static_cast<float>(y)) / HEIGHT) + h * 0.5f;
+
+		Native::Function::Call(Native::Hash::DRAW_RECT, xNew, yNew, w, h, color.R, color.G, color.B, color.A);
+	}
+	void ConsoleScript::DrawText(int x, int y, System::String ^text, float scale, int font, Color color)
```

Comment:
I'd make 'x' and 'y' floats. They are casted in the method anyway and this way we get rid of the "possible loss of data" warning in line 212 (since `GetTextLength` returns a float).